### PR TITLE
[FlagGems Operator Development Competition] Add smooth l1 loss

### PR DESCRIPTION
**Pull request by goldenfox2025**
## Summary

- Add Triton implementations for `aten.smooth_l1_loss` and `aten.smooth_l1_loss_backward`.
- Register `aten.smooth_l1_loss.out` and `aten.smooth_l1_loss_backward`.
- Add correctness tests for forward/backward reductions, beta values, broadcasting, non-contiguous tensors, special values, empty tensors, functional API, out overload, and invalid negative beta.
- Add dedicated benchmark shapes for both `smooth_l1_loss` and `smooth_l1_loss_backward`.

## Implementation Notes

Forward computes the standard Smooth L1 formula elementwise:

- `abs(input - target)` when `beta == 0`
- `0.5 * diff * diff / beta` when `abs(diff) < beta`
- `abs(diff) - 0.5 * beta` otherwise

The unreduced forward path writes the elementwise result directly. The reduced path writes fp32 partial sums and then performs a second-stage reduction, storing the final scalar in the requested output dtype.

Backward computes the input gradient using the Smooth L1 derivative and supports scalar or elementwise `grad_output`. For `reduction="mean"`, it matches PyTorch by dividing by the original input `numel()` before broadcasting. The `beta=0` and `input == target` edge case is matched to CUDA behavior; the CPU quick-mode test skips only that CUDA-specific regression because PyTorch CPU and CUDA differ there.

## Tests

```bash
export PATH="/usr/lib/wsl/lib:$PATH"
export GEMS_VENDOR=nvidia
export PYTHONPATH=src

python -m pytest -q tests/test_smooth_l1_loss.py
# 523 passed, 39 warnings

python -m pytest -s tests/test_smooth_l1_loss.py --ref=cpu --quick -x
# 522 passed, 1 skipped, 39 warnings

python -m pytest -q benchmark/test_smooth_l1_loss.py --query
# 2 passed, 38 warnings

python -m pytest benchmark/test_smooth_l1_loss.py -s --level core --warmup 20 --iter 50
# 2 passed, 39 warnings
```

## Test Coverage

- Operators: `smooth_l1_loss`, `smooth_l1_loss.out`, `smooth_l1_loss_backward`, and `torch.nn.functional.smooth_l1_loss`.
- Dtypes: fp16, fp32, bf16.
- Reductions: `none`, `mean`, `sum`.
- Beta values: `0.0`, `0.5`, `1.0`, `2.0`.
- Shapes: empty, scalar-like, 2D, 3D, 4D.
- Layout: non-contiguous input and non-contiguous backward `grad_output`.
- Broadcasting: forward `(2, 3, 4)` with `(4,)`; backward also covers `(2, 3, 4)` with `(4,)` and `(2, 1, 4)` with `(3, 4)`.
- Special values: `0`, `-0`, positive/negative values, `inf`, `-inf`, `nan`.
- Backward edge cases: scalar `grad_output`, CUDA `beta=0` equal-input behavior, and mean reduction after broadcasting.
- Invalid input: negative `beta`.

## Core Benchmark

Reduction is `mean`, beta is `1.0`.

### Forward

| Shape | fp16 speedup | fp32 speedup | bf16 speedup |
| --- | ---: | ---: | ---: |
| `(262144,)` | 1.531x | 1.300x | 1.357x |
| `(1024, 1024)` | 1.515x | 1.361x | 1.528x |
| `(4096, 4096)` | 1.627x | 1.933x | 1.674x |
| `(64, 512, 512)` | 1.610x | 1.920x | 1.636x |

### Backward

| Shape | fp16 speedup | fp32 speedup | bf16 speedup |
| --- | ---: | ---: | ---: |
| `(262144,)` | 1.667x | 1.368x | 1.813x |
| `(1024, 1024)` | 1.324x | 1.116x | 1.324x |
| `(4096, 4096)` | 1.342x | 1.356x | 1.338x |
| `(64, 512, 512)` | 1.363x | 2.015x | 1.345x |

### CODE REVIEW

**Review by bin913 — APPROVED** [review]

**Comment by bin913** [comment]
LGTM
